src/com/ms/silverking/cloud/dht/daemon/RingMapState.java

```diff
@@ -197,191 +194,195 @@ private void waitForExclusionSet() {
-            }
-            if (ringState != RingState.ABANDONED) {
-                Log.warning("Received notification: ready for convergence ", ringIDAndVersionPair);
-            } else {
-                Log.warning("Received notification: ABANDONED ", ringIDAndVersionPair);
-            }
-        }
+    newResolvedReplicaMap = newRingTree.getResolvedMap(ringConfig.getRingParentName(), new ReplicaNaiveIPPrioritizer());
+    ringTreeMinusExclusions = newRingTree;
+    resolvedReplicaMapMinusExclusions = newResolvedReplicaMap;
+    System.out.println("\tResolved Map");
+    resolvedReplicaMapMinusExclusions.display();
+
+    Log.warning("RingMapState done reading initial exclusions");
+  }
+
+  public void discard() {
+    synchronized (this) {
+      if (exclusionWatcher != null) {
+        exclusionWatcher.stop();
+      }
+      if (peerStateWatcher != null) {
+        peerStateWatcher.stop();
+      }
     }
-    
-    @Override
-    public void convergenceComplete(NamespaceConvergenceGroup ncg) {
-        transitionToState(RingState.LOCAL_CONVERGENCE_COMPLETE_1);
+  }
+
+  private void notifyReadyForConvergence() {
+    Log.warning("Notifying ready for convergence ", ringIDAndVersionPair);
+    synchronized (this) {
+      this.notifyAll();
     }
-    
-    private void transitionToState(RingState newRingState) {
-        if (verbosePeerStateTransition) {
-            Log.warning("transitionToState: "+ newRingState);
-        }
-        if (ringState == null || ringState.isValidTransition(newRingState)) {
-            ringState = newRingState;
-            localStateActions(newRingState);
-            try {
-                ringStateZK.writeState(nodeID, ringState);
-            } catch (KeeperException ke) {
-                throw new RuntimeException("handle", ke); // FIXME - handle as fatal for this update
-            }
-        } else {
-            throw new InvalidTransitionException(ringState, newRingState);
-        }
+  }
+
+  public void abandonConvergence() {
+    synchronized (this) {
+      Log.warningf("RingMapState.abandonConvergence() %s", ringIDAndVersionPair);
+      peerStateWatcher.stop();
+      if (!ringState.isFinal()) {
+        Log.warning("Notifying abandoned ", ringIDAndVersionPair);
+        transitionToState(RingState.ABANDONED);
+        this.notifyAll();
+      }
     }
-    
-    public TransitionReplicaSources getReplicaSources(RingOwnerQueryOpType opType) {
-        switch (opType) {
-        case Write: return writeTargets;
-        case Read: return readTargets;
-        default: throw new RuntimeException("panic");
+  }
+
+  public void waitUntilReadyForConvergence() {
+    synchronized (this) {
+      Log.warning("Waiting until ready for convergence ", ringIDAndVersionPair);
+      while (ringState == RingState.INITIAL || ringState == RingState.READY_FOR_CONVERGENCE_1) {
+        try {
+          Log.warning(ringState);
+          this.wait();
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)